### PR TITLE
Add runLayout helper to run one of the available layouts in the current available graph

### PR DESCRIPTION
**Pull request by tarcisiofischer**

### CODE REVIEW

File: qmxgraph/page/api.js [hunk 1]
```diff
@@ -1646,3 +1646,41 @@ graphs.Api.prototype._findDecoration = function _findDecoration (cellId) {
...
+
+/**
+ * Automatically organize the graph using one of the available layouts. If the layout is not
+ * implemented, an exception will be raised.
+ *
+ * @param {string} layoutName The layout name (e.g.: 'organic'). Check the code for the available
+ * layouts.
+ * @throws {Error} If the layout is unknown.
+ */
+graphs.Api.prototype.runLayout = function runLayout(layoutName) {
```
Comment (nicoddemus) [inline qmxgraph/page/api.js]: Can we have at least a shallow test to ensure this is not buggy?

File: tests/test_js_graph.py [hunk 2]
```diff
@@ -1502,3 +1503,28 @@ def insert_by_parametrized_type(graph, cell_type, tags=None):
...
+        QmxGraphApi.LAYOUT_CIRCLE,
+        QmxGraphApi.LAYOUT_COMPACT_TREE,
+        QmxGraphApi.LAYOUT_EDGE_LABEL,
+        QmxGraphApi.LAYOUT_PARALLEL_EDGE,
+        QmxGraphApi.LAYOUT_PARTITION,
+        QmxGraphApi.LAYOUT_RADIAL_TREE,
+        QmxGraphApi.LAYOUT_STACK,
+    ],
+)
+def test_run_all_layouts(layout_name, graph_cases):
```
Comment (tarcisiofischer) [inline tests/test_js_graph.py]: I tried verifying the position before and after the run, but there are cases where they can be the same... I don't know a good way to test this :(

Leaving the (bad) test here for reference:

```python
@pytest.mark.parametrize(
    'layout_name',
    [
        QmxGraphApi.LAYOUT_ORGANIC,
        QmxGraphApi.LAYOUT_COMPACT,
        QmxGraphApi.LAYOUT_CIRCLE,
        QmxGraphApi.LAYOUT_COMPACT_TREE,
        QmxGraphApi.LAYOUT_EDGE_LABEL,
        QmxGraphApi.LAYOUT_PARALLEL_EDGE,
        QmxGraphApi.LAYOUT_PARTITION,
        QmxGraphApi.LAYOUT_RADIAL_TREE,
        QmxGraphApi.LAYOUT_STACK,
    ],
)
def test_run_all_layouts(layout_name, graph_cases):
    graph = graph_cases('3v_1e')
    label = lambda cell: graph.get_label(cell)
    nodes_positions = {
        label(v): {
            'before': None,
            'after': None,
        }
        for v in graph.get_vertices()
    }

    for v in graph.get_vertices():
        nodes_positions[label(v)]['before'] = graph.get_vertex_position(v)
    graph.eval_js_function('api.runLayout', layout_name)
    for v in graph.get_vertices():
        nodes_positions[label(v)]['after'] = graph.get_vertex_position(v)

    for position_data in nodes_positions.values():
        # We do not have the exact expected position to check - But we do know that the positions
        # should at least change.
        x_before, y_before = position_data['before']
        x_after, y_after = position_data['after']
        assert not pytest.approx(x_before) == x_after
        assert not pytest.approx(y_before) == y_after

Comment (nicoddemus) [reply tests/test_js_graph.py]: Not sure how to proceed here as well...

Comment (prusse-martin) [reply tests/test_js_graph.py]: A single test checking if organic layout works? (do you know others?)

How does the layout algorithms work with the decorations and/or tables?
Are they ignored? Moved?
What if the cell are relativelly positioned?

Comment (tarcisiofischer) [reply tests/test_js_graph.py]: **A single test checking if organic layout works? (do you know others?)**
Even the organic layout can return some cells in the same position, and it is stochastic (Results will change in each run). How can I check the results?

**How does the layout algorithms work with the decorations and/or tables? Are they ignored? Moved?**
We don't really care about decorations and tables... I didn't tested, but I don't care if they are moved around.

**What if the cell are relativelly positioned?**
I didn't even know this is possible. How can this be checked?

Comment (prusse-martin) [reply tests/test_js_graph.py]: Ah... also note that `3v_1e` has just 1 edge having an disconnected vertex and disconected vertices are not moved.

Comment (prusse-martin) [reply tests/test_js_graph.py]: > (Results will change in each run). How can I check the results?
In you test you could mock the `Math.random` to return a fixed value yielding deterministic results.

> I didn't even know this is possible. How can this be checked?
Maybe `cell.getGeometry().relative`;

Comment (tarcisiofischer) [reply tests/test_js_graph.py]: How do I mock a Javascript function from the python test?

File: tests/test_js_graph.py [hunk 3]
```diff
@@ -398,7 +399,7 @@ def test_get_geometry_plain(graph_cases):
     assert graph.get_geometry(graph.get_edge(*graph.get_vertices())) == \
         [40, 25, 50, 1]
     assert graph.get_geometry(graph.get_decorations()[0]) == [55, 20, 10, 10]
-    table_w, table_h = fix_table_size(102, 72)
+    table_w, table_h = fix_table_size(100, 72)
```
Comment (prusse-martin) [inline tests/test_js_graph.py]: I created #58 to find a way to better handle  `fix_table_size`.
=/ It is really anoying having to tune they areguments from time to time. Is it not?

File: tests/test_js_graph.py (hunk 2 above)
Comment (prusse-martin) [reply tests/test_js_graph.py]: There is not a implementation for that I think.

```python
def mocked_js_random(graph):
    graph.selenium.execute_script("_mocked_js_random_original = Math.random;")
    # 4: Chosen by fair dice roll (d6). Guaranteed to be random.
    # Math.random => [0, 1)
    # (4 - 1) / 6 = 0.5
    graph.selenium.execute_script("Math.random = function(){ return 0.5; }")
    try:
        yield
    finally:
        graph.selenium.execute_script("Math.random = _mocked_js_random_original;")
```

File: tests/test_js_graph.py [hunk 4]
```diff
@@ -1502,3 +1503,54 @@ def insert_by_parametrized_type(graph, cell_type, tags=None):
...
+    graph.eval_js_function('api.runLayout', QmxGraphApi.LAYOUT_ORGANIC)
+    for v in graph.get_vertices():
+        nodes_positions[label(v)]['after'] = graph.get_vertex_position(v)
+
+    for position_data in nodes_positions.values():
+        # We do not have the exact expected position to check - But we do know that the positions
+        # should at least change.
+        x_before, y_before = position_data['before']
+        x_after, y_after = position_data['after']
+        assert not pytest.approx(x_before) == x_after
```
Comment (prusse-martin) [inline tests/test_js_graph.py]: Add the `position_data` the the assert message plz.

File: tests/test_js_graph.py (hunk 3 above)
Comment (tarcisiofischer) [reply tests/test_js_graph.py]: I don't know why this changed, doesn't seem related to my changes.
I don't really cared about it

Comment (prusse-martin) [reply tests/test_js_graph.py]: Yes... this is kind of flaky.

File: tests/test_js_graph.py [hunk 5]
```diff
@@ -1502,3 +1503,54 @@ def insert_by_parametrized_type(graph, cell_type, tags=None):
...
+    graph.eval_js_function('api.runLayout', QmxGraphApi.LAYOUT_ORGANIC)
+    for v in graph.get_vertices():
+        nodes_positions[label(v)]['after'] = graph.get_vertex_position(v)
+
+    for position_data in nodes_positions.values():
+        # We do not have the exact expected position to check - But we do know that the positions
+        # should at least change.
+        x_before, y_before = position_data['before']
+        x_after, y_after = position_data['after']
+        assert not pytest.approx(x_before) == x_after and not pytest.approx(y_before) == y_after, \
```
Comment (prusse-martin) [inline tests/test_js_graph.py]: Also I guess you could do `assert not pytest.approx(position_data['before']) ==position_data['after'], '...'`.
But you don't need to change that =)

**Comment by prusse-martin** [comment]
Mastaer is also having a timeout in 3.6.
I will suguest to cahnge the string formating to explictly call `.format`(instead using f-strings) so we got 3.5 passing and we could merge this.
(BTW, sorry for noticing this half a day after having approved the pr contents)

**Comment by prusse-martin** [comment]
My changes should resolve #58 (since there is no longer a `fix_table_size` =)

**Pull request merged**